### PR TITLE
Cherry pick some commits from appleseedX

### PR DESCRIPTION
**Pull request by est77**

### CODE REVIEW

File: src/appleseed/foundation/cuda/cudadevice.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+// Interface header.
+#include "cudadevice.h"
+
+// appleseed.foundation headers.
+#include "foundation/cuda/exception.h"
+
+// Standard headers.
+#include <cassert>
+#include <algorithm>
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/cudadevice.cpp]: Order includes

File: src/appleseed/CMakeLists.txt [hunk 2]
```diff
@@ -107,6 +107,22 @@ source_group ("foundation\\core" FILES
...
+    set (foundation_cuda_sources
+        foundation/cuda/cudadevice.cpp
+        foundation/cuda/cudadevice.h
+        foundation/cuda/exception.cpp
+        foundation/cuda/exception.h
+    )
+
+    list (APPEND appleseed_sources
+        ${foundation_cuda_sources}
+    )
```
Comment (oktomus) [inline src/appleseed/CMakeLists.txt]: Blank line after this

File: src/appleseed/foundation/cuda/cudadevice.h [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#ifndef APPLESEED_FOUNDATION_CUDA_CUDADEVICE_H
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/cudadevice.h]: #pragma only_once ?

File: src/appleseed/foundation/cuda/exception.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#ifndef APPLESEED_FOUNDATION_CORE_CUDA_EXCEPTION_H
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/exception.h]: only_once also here

**Comment by est77** [comment]
> It's really nice that you are bringing CUDA into appleseed.

For now it is only cherry picking some commits from the appleseedX branch and merging them before they get too much out of sync.

File: src/appleseed/foundation/platform/system.cpp [hunk 5]
```diff
@@ -952,6 +1025,10 @@ void System::print_information(Logger& logger)
         pretty_size(get_total_virtual_memory_size()).c_str(),
         pretty_uint(DefaultWallclockTimer().frequency()).c_str(),
         pretty_uint(DefaultProcessorTimer().frequency()).c_str());
+
+#ifdef APPLESEED_WITH_GPU
+    print_gpu_information(logger);
```
Comment (oktomus) [inline src/appleseed/foundation/platform/system.cpp]: Indent is not correct, missing 4 spaces

File: src/appleseed/foundation/platform/system.cpp [hunk 6]
```diff
@@ -95,6 +99,10 @@
 
 #endif
 
+#ifdef APPLESEED_WITH_GPU
+    #include <optix.h>
```
Comment (oktomus) [inline src/appleseed/foundation/platform/system.cpp]: Remove tab here (to be consistent with includes line 34.

File: src/appleseed/foundation/cuda/exception.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// appleseed.foundation headers.
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/exception.cpp]: "Interface header."

**Comment by est77** [comment]
updated and added one more commit

File: sandbox/kernels/.gitignore [hunk 8]
```diff
@@ -0,0 +1 @@
+*
```
Comment (oktomus) [inline sandbox/kernels/.gitignore]: Is it necessary ? If what you want to achieve is ignore everything in this folder, you should ignore `sandbox/kernels` using your local conf.

Comment (oktomus) [reply sandbox/kernels/.gitignore]: Oh, CUDA builds output to this directory and this is why you ignore it. Forget what I said.

File: src/appleseed/foundation/cuda/cudadevice.h [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    // Constructor.
+    explicit CUDADevice(const CUdevice device_number);
+
+    // Device properties.
+    CUdevice            m_cuda_device_number;
+    APIString           m_name;
+    std::pair<int, int> m_compute_capability;
+    int                 m_compute_mode;
+
+    size_t              m_total_mem;
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/cudadevice.h]: It should be `std::size_t` and `#include <cstddef>` should be added.

File: src/appleseed/foundation/cuda/cudadevice.h [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+  public:
+    static const CUDADeviceList& instance();
+
+    ~CUDADeviceList();
+
+    bool empty() const;
+
+    std::size_t size() const;
+
+    const CUDADevice& get_device(const size_t device_number) const;
```
Comment (oktomus) [inline src/appleseed/foundation/cuda/cudadevice.h]: `std::stize_t`

File: src/appleseed/foundation/platform/system.cpp [hunk 11]
```diff
@@ -95,6 +99,13 @@
 
 #endif
 
+#ifdef APPLESEED_WITH_GPU
```
Comment (oktomus) [inline src/appleseed/foundation/platform/system.cpp]: In the block above, you haven't added blank lines:
```
#ifdef APPLESEED_WITH_GPU
#include "foundation/cuda/cudadevice.h"
#include "foundation/cuda/exception.h"
#endif
```

You should do the same here.

Comment (est77) [reply src/appleseed/foundation/platform/system.cpp]: I prefer to be consistent with the per OS / system blocks, just because they are closer.

**Pull request merged**